design/httpproxy-conflict.md

```diff
@@ -0,0 +1,123 @@
+These errors can come from invalid route configuration, missing service references, or invalid includes to other resources.
+To fix this, Contour will treat errors and warning differently and where possible, make a best effort to serve partially valid configurations.
+Configurations that are in an Error or Warning status, will serve HTTP error codes or not be added to configuration passed down to Envoy.
+
+## Personas
+
+There are two main personas that interact with this conflict design document:
+
+- User: A request to an application, routing through Envoy looking to access whatever resource an application running in Kubernetes should serve. This user is not aware of how the application is deployed or that it uses Kubernetes or Contour.
+- ResourceOwner: Someone to manages an HTTPProxy resource inside the Kubernetes cluster configuring the Route/Service combination to allow users of the application to access.
+
+## Background
+HTTPProxy is designed as a multi-team aware Ingress that reduces annotation sprawl by providing a sensible home for configuration items.
+Additionally, HTTPProxy offers a feedback loop to ResourceOwners (See #Personas section below) via a `Status` field on the object.
+Contour has further extended this to add `Conditions` which allows for a set of `Errors` and `Warnings` which inform users of errors in the HTTPProxy resource.
+
+It's difficult to find errors and block them from breaking valid configurations since Contour currently processes objects after they are committed to the API server.
+Due to this, Contour cannot block changes to resources before they are committed like an `Admission Controller` would be able to do (See alternatives: `HTTPProxyInstance`).
+When processing HTTPProxy objects, some problems should not stop the rest of the object from being processed.
+To distinguish between fatal-for-processing and not, we propose calling the former `Errors`, and the latter `Warnings`.
+
+When the first error or warning is detected, these fields are updated and in the status of the HTTPProxy object, however if there are more errors, only the first is shown to the user.
+There are ongoing efforts to change this logic to inform of all the errors or warnings dealing with a specific object.
+
+Contour currently invalidates the entire resource if a configuration error is encountered which can cause downtime for specific routes or even the entire virtual host.
+
+When processing HTTPProxy objects, some problems should not stop the rest of the object from being processed.
+Contour will set an object to be status `Error` if the request response to the `User` is changed from what is configured.
+Contour will set a `Warning` when the object has an issue, but the response a `User` sees when accessing the application for which the resource is configured is not modified.
+
+## Goals
+- Serve valid traffic configurations if portions of the spec are invalid
+- Define and distinguish between fatal and non-fatal errors
```

Comment:
Since it's a goal to define these, I think it'd be great to add a short **Definitions** section that very precisely defines each of these, something like (feel free to adjust):

## Definitions
### Fatal Error
An error encountered while processing an HTTPProxy spec where:
- Contour stops processing the rest of the spec
- Contour does not program any part of the HTTPProxy in Envoy

For example, an invalid FQDN.

### Non-Fatal Error
An error encountered while processing an HTTPProxy spec where:
- Contour continues processing the rest of the spec
- Contour *does* program the unaffected part of the HTTPProxy in Envoy
- Users may get a different response for a given route than what was configured (e.g. a 502 instead of the desired upstream service)

For example, a route to a nonexistent service (results in a 502 being programmed)

### Warning
An error encountered while processing an HTTPProxy spec where:
- Contour continues processing the rest of the spec
- Contour *does* program the unaffected part of the HTTPProxy in Envoy
- Users are routed to the correct destinations, but some aspects of the request handling may be different than desired (e.g. missing timeout or retry settings)

For example, invalid retry settings on a route (the route is still programmed, but without retry settings)